File: backend/app/observability/sentry.py
"""Sentry error monitoring integration for Batchivo backend.

Provides centralized error tracking, performance monitoring, and release tracking
with Sentry. Integrates with FastAPI ASGI middleware for automatic request tracing.
"""

import logging
import subprocess

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
from sentry_sdk.integrations.redis import RedisIntegration
from sentry_sdk.integrations.celery import CeleryIntegration
from sentry_sdk.integrations.logging import LoggingIntegration

from app.config import get_settings

logger = logging.getLogger(__name__)


def get_git_release() -> str | None:
    """Get the current git commit hash for release tracking."""
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--short", "HEAD"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        if result.returncode == 0:
            return f"batchivo-backend@{result.stdout.strip()}"
    except Exception:
        pass
    return None


def init_sentry() -> bool:
    """
    Initialize Sentry SDK for error monitoring.

    Returns:
        bool: True if Sentry was initialized successfully, False otherwise.
    """
    settings = get_settings()

    # Check if Sentry DSN is configured
    sentry_dsn = getattr(settings, "sentry_dsn", None)
    if not sentry_dsn:
        logger.info("Sentry DSN not configured, skipping initialization")
        return False

    # Get release version from git or settings
    release = get_git_release()

    # Configure sample rates based on environment
    traces_sample_rate = 0.1  # 10% in production
    profiles_sample_rate = 0.1

    if settings.environment == "development":
        traces_sample_rate = 1.0  # 100% in development
        profiles_sample_rate = 1.0
    elif settings.environment == "staging":
        traces_sample_rate = 0.5  # 50% in staging
        profiles_sample_rate = 0.5

    try:
        sentry_sdk.init(
            dsn=sentry_dsn,
            environment=settings.environment,
            release=release,
            traces_sample_rate=traces_sample_rate,
            profiles_sample_rate=profiles_sample_rate,
            # Send PII only in development for debugging
            send_default_pii=settings.is_development,
            # Attach request data to events
            request_bodies="medium",
            # Enable performance monitoring
            enable_tracing=True,
            # Integrations
            integrations=[
                # FastAPI and Starlette for request/response tracing
                StarletteIntegration(
                    transaction_style="endpoint",
                ),
                FastApiIntegration(
                    transaction_style="endpoint",
                ),
                # SQLAlchemy for database query tracing
                SqlalchemyIntegration(),
                # Redis for cache operation tracing
                RedisIntegration(),
                # Celery for background job tracing
                CeleryIntegration(),
                # Logging integration - capture errors from logging
                LoggingIntegration(
                    level=logging.INFO,  # Capture INFO and above
                    event_level=logging.ERROR,  # Only send ERROR and above as events
                ),
            ],
            # Filter out health check endpoints from performance data
            traces_sampler=traces_sampler,
            # Add custom tags
            before_send=before_send,
        )

        logger.info(f"Sentry initialized (env={settings.environment}, release={release})")
        return True

    except Exception as e:
        logger.error(f"Failed to initialize Sentry: {e}")
        return False


def traces_sampler(sampling_context: dict) -> float:
    """
    Custom traces sampler to filter out noisy endpoints.

    Args:
        sampling_context: Context about the transaction being sampled.

    Returns:
        Sample rate (0.0 to 1.0) or None to use default rate.
    """
    settings = get_settings()

    # Get transaction name if available
    transaction_context = sampling_context.get("transaction_context", {})
    name = transaction_context.get("name", "")

    # Never trace health check endpoints
    if any(path in name for path in ["/health", "/ready", "/live", "/metrics"]):
        return 0.0

    # Lower sample rate for high-volume endpoints
    if "/api/v1/dashboard" in name or "/api/v1/analytics" in name:
        return 0.05  # 5% for dashboard/analytics

    # Use environment-based defaults
    if settings.environment == "development":
        return 1.0
    elif settings.environment == "staging":
        return 0.5
    else:
        return 0.1


def before_send(event: dict, hint: dict) -> dict | None:
    """
    Process events before sending to Sentry.

    Can be used to filter, modify, or enrich events.

    Args:
        event: The Sentry event dictionary.
        hint: Additional context about the event.

    Returns:
        Modified event dict, or None to drop the event.
    """
    # Don't send events for expected/handled exceptions
    if "exc_info" in hint:
        exc_type, exc_value, exc_tb = hint["exc_info"]

        # Don't report 404s or rate limit errors as Sentry issues
        from fastapi import HTTPException
        from slowapi.errors import RateLimitExceeded

        if isinstance(exc_value, HTTPException):
            if exc_value.status_code in [404, 401, 403, 429]:
                return None

        if isinstance(exc_value, RateLimitExceeded):
            return None

    # Add custom context
    settings = get_settings()
    event.setdefault("tags", {})
    event["tags"]["app.name"] = settings.app_name
    event["tags"]["storage.type"] = settings.storage_type

    return event


def capture_exception(error: Exception, **extra) -> str | None:
    """
    Manually capture an exception and send to Sentry.

    Args:
        error: The exception to capture.
        **extra: Additional context to attach to the event.

    Returns:
        Event ID if captured, None otherwise.
    """
    with sentry_sdk.push_scope() as scope:
        for key, value in extra.items():
            scope.set_extra(key, value)
        return sentry_sdk.capture_exception(error)


def capture_message(message: str, level: str = "info", **extra) -> str | None:
    """
    Capture a message and send to Sentry.

    Args:
        message: The message to capture.
        level: Severity level (debug, info, warning, error, fatal).
        **extra: Additional context to attach to the event.

    Returns:
        Event ID if captured, None otherwise.
    """
    with sentry_sdk.push_scope() as scope:
        for key, value in extra.items():
            scope.set_extra(key, value)
        return sentry_sdk.capture_message(message, level=level)


def set_user(user_id: str | None, email: str | None = None, **extra) -> None:
    """
    Set user context for Sentry events.

    Call this when a user is authenticated to associate errors with users.

    Args:
        user_id: The user's unique identifier.
        email: The user's email address.
        **extra: Additional user attributes.
    """
    user_data = {"id": user_id}
    if email:
        user_data["email"] = email
    user_data.update(extra)
    sentry_sdk.set_user(user_data)


def set_tenant(tenant_id: str, tenant_name: str | None = None) -> None:
    """
    Set tenant context for Sentry events.

    Call this when tenant context is established to associate errors with tenants.

    Args:
        tenant_id: The tenant's unique identifier.
        tenant_name: The tenant's display name.
    """
    sentry_sdk.set_tag("tenant.id", tenant_id)
    if tenant_name:
        sentry_sdk.set_tag("tenant.name", tenant_name)

File: frontend/src/lib/sentry.ts
/**
 * Sentry Error Monitoring for Batchivo Frontend
 *
 * Provides error tracking, performance monitoring, and session replay
 * with proper integration into React and the existing ErrorBoundary.
 */

import * as Sentry from '@sentry/react'
import { config } from './config'

let isInitialized = false

/**
 * Get git commit hash for release tracking.
 * This is set at build time by the Vite plugin.
 */
function getRelease(): string {
  // __SENTRY_RELEASE__ is injected by @sentry/vite-plugin
  // Falls back to app version if not available
  return (
    (typeof __SENTRY_RELEASE__ !== 'undefined' ? __SENTRY_RELEASE__ : null) ||
    `batchivo-frontend@${config.serviceVersion}`
  )
}

/**
 * Initialize Sentry error monitoring.
 * Should be called once at application startup, before React renders.
 *
 * @returns true if Sentry was initialized, false if skipped
 */
export function initSentry(): boolean {
  if (isInitialized) {
    console.warn('Sentry already initialized')
    return true
  }

  // Skip if no DSN configured
  if (!config.sentryDsn) {
    console.info('Sentry DSN not configured, skipping initialization')
    return false
  }

  try {
    // Configure sample rates based on environment
    const tracesSampleRate = config.isDev ? 1.0 : 0.1
    const replaysSessionSampleRate = config.isDev ? 1.0 : 0.1
    const replaysOnErrorSampleRate = 1.0 // Always capture replay on error

    Sentry.init({
      dsn: config.sentryDsn,
      environment: config.mode,
      release: getRelease(),

      // Performance Monitoring
      tracesSampleRate,

      // Session Replay
      replaysSessionSampleRate,
      replaysOnErrorSampleRate,

      // Integrations
      integrations: [
        // Browser tracing for performance
        Sentry.browserTracingIntegration({
          // Trace navigation and API calls
          tracePropagationTargets: [
            'localhost',
            /^https:\/\/.*\.batchivo\.com/,
            /^\/api\//,
          ],
        }),
        // Session replay for debugging
        Sentry.replayIntegration({
          // Mask all text and block all media for privacy in production
          maskAllText: config.isProd,
          blockAllMedia: config.isProd,
        }),
      ],

      // Filter events before sending
      beforeSend(event, hint) {
        // Don't send events in development unless explicitly enabled
        if (config.isDev && !import.meta.env.VITE_SENTRY_DEV_ENABLED) {
          console.debug('Sentry event (dev mode, not sent):', event)
          return null
        }

        // Filter out non-error events for common issues
        const error = hint.originalException
        if (error instanceof Error) {
          // Don't report network errors (handled by error boundaries)
          if (
            error.message.includes('Network request failed') ||
            error.message.includes('Failed to fetch')
          ) {
            return null
          }

          // Don't report ResizeObserver errors (browser quirk)
          if (error.message.includes('ResizeObserver')) {
            return null
          }
        }

        return event
      },

      // Filter breadcrumbs
      beforeBreadcrumb(breadcrumb) {
        // Don't log console breadcrumbs in production
        if (config.isProd && breadcrumb.category === 'console') {
          return null
        }
        return breadcrumb
      },

      // Additional options
      sendDefaultPii: config.isDev, // Only send PII in development
      attachStacktrace: true,
      autoSessionTracking: true,
    })

    // Set initial tags
    Sentry.setTag('app.name', 'batchivo-frontend')
    Sentry.setTag('app.version', config.serviceVersion)

    isInitialized = true
    console.info(`Sentry initialized (env=${config.mode}, release=${getRelease()})`)
    return true
  } catch (error) {
    console.error('Failed to initialize Sentry:', error)
    return false
  }
}

/**
 * Set user context for Sentry events.
 * Call when user logs in.
 */
export function setUser(user: {
  id: string
  email?: string
  username?: string
  tenantId?: string
}): void {
  Sentry.setUser({
    id: user.id,
    email: user.email,
    username: user.username,
  })

  if (user.tenantId) {
    Sentry.setTag('tenant.id', user.tenantId)
  }
}

/**
 * Clear user context.
 * Call when user logs out.
 */
export function clearUser(): void {
  Sentry.setUser(null)
}

/**
 * Capture an exception manually.
 */
export function captureException(
  error: Error | unknown,
  context?: Record<string, unknown>
): string | undefined {
  return Sentry.captureException(error, {
    extra: context,
  })
}

/**
 * Capture a message manually.
 */
export function captureMessage(
  message: string,
  level: 'fatal' | 'error' | 'warning' | 'info' | 'debug' = 'info',
  context?: Record<string, unknown>
): string | undefined {
  return Sentry.captureMessage(message, {
    level,
    extra: context,
  })
}

/**
 * Add breadcrumb for debugging.
 */
export function addBreadcrumb(
  message: string,
  category: string,
  level: 'fatal' | 'error' | 'warning' | 'info' | 'debug' = 'info',
  data?: Record<string, unknown>
): void {
  Sentry.addBreadcrumb({
    message,
    category,
    level,
    data,
  })
}

/**
 * Create a span for performance monitoring.
 */
export function startSpan<T>(
  name: string,
  callback: () => T | Promise<T>,
  options?: {
    op?: string
    description?: string
    attributes?: Record<string, string | number | boolean>
  }
): T | Promise<T> {
  return Sentry.startSpan(
    {
      name,
      op: options?.op || 'function',
      attributes: options?.attributes,
    },
    callback
  )
}

// Re-export ErrorBoundary from Sentry for use in App
export const SentryErrorBoundary = Sentry.ErrorBoundary

// Declare global for release injection
declare global {
  const __SENTRY_RELEASE__: string | undefined
}
